pkg/controllers/subnet/subnet_controller.go

```diff
@@ -50,102 +48,144 @@ type SubnetReconciler struct {
-			deleteSuccess(r, ctx, obj)
-		} else {
-			log.Info("finalizers cannot be recognized", "subnet", req.NamespacedName)
+		log.Info("Updated Subnet CR", "Subnet", req.NamespacedName)
+	}
+
+	tags := r.SubnetService.GenerateSubnetNSTags(subnetCR)
+	if tags == nil {
+		log.Error(nil, "failed to generate Subnet tags", "Subnet", req.NamespacedName)
+		return ResultRequeue, errors.New("failed to generate Subnet tags")
+	}
+	// List VPC Info
+	vpcInfoList := r.VPCService.ListVPCInfo(req.Namespace)
+	if len(vpcInfoList) == 0 {
+		log.Info("No VPC info found, requeueing", "Namespace", req.Namespace)
+		return ResultRequeueAfter10sec, nil
+	}
+	// Create or update the subnet in NSX
+	if _, err := r.SubnetService.CreateOrUpdateSubnet(subnetCR, vpcInfoList[0], tags); err != nil {
+		if errors.As(err, &nsxutil.ExceedTagsError{}) {
+			log.Error(err, "tags limit exceeded, not retrying", "Subnet", req.NamespacedName)
+			updateFail(r, ctx, subnetCR, err.Error())
+			return ResultNormal, nil
 		}
+		log.Error(err, "failed to create/update Subnet, retrying", "Subnet", req.NamespacedName)
+		updateFail(r, ctx, subnetCR, err.Error())
+		return ResultRequeue, err
 	}
+	// Update status
+	if err := r.updateSubnetStatus(subnetCR); err != nil {
+		log.Error(err, "failed to update Subnet status, retrying", "Subnet", req.NamespacedName)
+		updateFail(r, ctx, subnetCR, err.Error())
+		return ResultRequeue, err
+	}
+	updateSuccess(r, ctx, subnetCR)
 	return ctrl.Result{}, nil
 }
 
-func (r *SubnetReconciler) DeleteSubnet(obj v1alpha1.Subnet) error {
-	nsxSubnets := r.SubnetService.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetCRUID, string(obj.GetUID()))
-	if len(nsxSubnets) == 0 {
-		log.Info("no subnet found for subnet CR", "uid", string(obj.GetUID()))
-		return nil
+func (r *SubnetReconciler) deleteSubnetByID(subnetID string) error {
+	nsxSubnets := r.SubnetService.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetCRUID, subnetID)
+	return r.deleteSubnets(nsxSubnets)
+}
+
+func (r *SubnetReconciler) deleteSubnets(nsxSubnets []*model.VpcSubnet) error {
+	for _, nsxSubnet := range nsxSubnets {
+		portNums := len(r.SubnetPortService.GetPortsOfSubnet(*nsxSubnet.Id))
+		if portNums > 0 {
+			err := errors.New("subnet still attached by port")
+			log.Error(err, "delete Subnet from NSX failed", "ID", *nsxSubnet.Id)
+			return err
+		}
+		if err := r.SubnetService.DeleteSubnet(*nsxSubnet); err != nil {
+			log.Error(err, "failed to delete Subnet", "ID", *nsxSubnet.Id)
+			return err
+		}
 	}
-	portNums := len(r.SubnetPortService.GetPortsOfSubnet(*nsxSubnets[0].Id))
-	if portNums > 0 {
-		err := errors.New("subnet still attached by port")
-		log.Error(err, "", "ID", *nsxSubnets[0].Id)
-		return err
+	return nil
+}
+
+func (r *SubnetReconciler) deleteSubnetByName(name, ns string) error {
+	nsxSubnets := r.SubnetService.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetCRNamespacedName, util.CombineNamespaceName(name, ns))
```

Comment:
for the nsxSubnets gotten from store, we need to also get it's K8s uid to check if this CR is deleted or not, if CR uid is not in k8s, then we go to delete nsx subnet. if K8s uid CR exists, we skip to delete.

Reply:
Updated.